### PR TITLE
Implementation of realising XTalk for Inner Pixel detector (IT)

### PR DESCRIPTION
**Pull request by suchandradutta**
This PR introduces a realistic description of the inter-pixel cross-talk in the InnerTracker for 25x100 um^2 pixel cells (phase2 only) as determined by lab measurements on prototype chip+sensor assemblies. It affects the phase2 work flows with the following versions of the Tracker geometry: T5, T6, T7, T8, T9, T10. At the moment there is no evidence of x-talk in other pixel sizes. The work has been presented in Tracker Phase2 Simulation Meeting [*

[*]https://indico.cern.ch/event/768246/contributions/3199413/attachments/1746858/2828470/CMSSW_inter-pixel_cross-talk.pdf

@emiglior 
@ajofrehe 
@Arash-Jofrehei

### CODE REVIEW

**Comment by boudoul** [comment]
Hi @suchandradutta , could you please apply the patch which is sggested by the bot , it is a trivial change - Thanks

File: SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizerAlgorithm.cc [hunk 1]
```diff
@@ -592,62 +592,75 @@ void Phase2TrackerDigitizerAlgorithm::induce_signal(const PSimHit& hit,
...
+void Phase2TrackerDigitizerAlgorithm::add_cross_talk(const Phase2TrackerGeomDetUnit* pixdet) {
+  uint32_t detID = pixdet->geographicalId().rawId();
+  signal_map_type& theSignal = _signal[detID];
+  signal_map_type signalNew;
+  const Phase2TrackerTopology* topol = &pixdet->specificTopology();
+  int numRows = topol->nrows();
+
+  for (auto & s : theSignal) {
+    float signalInElectrons = s.second.ampl();   // signal in electrons
+    std::pair<int,int> hitChan;
```
Comment (kpedro88) [inline SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizerAlgorithm.cc]: `auto hitChan = (pixelFlag) ? PixelDigi::channelToPixel(s.first) : Phase2TrackerDigi::channelToPixel(s.first);`

File: SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizerAlgorithm.cc [hunk 2]
```diff
@@ -592,62 +592,75 @@ void Phase2TrackerDigitizerAlgorithm::induce_signal(const PSimHit& hit,
...
+    if (pixelFlag) hitChan = PixelDigi::channelToPixel(s.first);
+    else hitChan = Phase2TrackerDigi::channelToPixel(s.first);
+    
+    float signalInElectrons_Xtalk = signalInElectrons * interstripCoupling;     
+    //subtract the charge which will be shared 
+    s.second.set(signalInElectrons-signalInElectrons_Xtalk);
+         
+    if (hitChan.first != 0) {
+      std::pair<int,int> XtalkPrev = std::pair<int,int>(hitChan.first-1, hitChan.second);
```
Comment (kpedro88) [inline SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizerAlgorithm.cc]: `auto XtalkPrev = std::make_pair(hitChan.first-1, hitChan.second);`

File: SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizerAlgorithm.cc [hunk 3]
```diff
@@ -592,62 +592,75 @@ void Phase2TrackerDigitizerAlgorithm::induce_signal(const PSimHit& hit,
...
+      int chanXtalkPrev = (pixelFlag) ? PixelDigi::pixelToChannel(XtalkPrev.first, XtalkPrev.second)
+	: Phase2TrackerDigi::pixelToChannel(XtalkPrev.first, XtalkPrev.second);
+      signalNew.insert(std::pair<int,DigitizerUtility::Amplitude>(chanXtalkPrev, DigitizerUtility::Amplitude(signalInElectrons_Xtalk, nullptr, -1.0)));
+    }
+    if (hitChan.first < (numRows-1)) {
+      std::pair<int,int> XtalkNext = std::pair<int,int>(hitChan.first+1, hitChan.second);
```
Comment (kpedro88) [inline SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizerAlgorithm.cc]: `auto XtalkNext = std::make_pair(hitChan.first+1, hitChan.second);`

File: SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizerAlgorithm.cc [hunk 4]
```diff
@@ -592,62 +592,75 @@ void Phase2TrackerDigitizerAlgorithm::induce_signal(const PSimHit& hit,
...
+    
+    float signalInElectrons_Xtalk = signalInElectrons * interstripCoupling;     
+    //subtract the charge which will be shared 
+    s.second.set(signalInElectrons-signalInElectrons_Xtalk);
+         
+    if (hitChan.first != 0) {
+      std::pair<int,int> XtalkPrev = std::pair<int,int>(hitChan.first-1, hitChan.second);
+      int chanXtalkPrev = (pixelFlag) ? PixelDigi::pixelToChannel(XtalkPrev.first, XtalkPrev.second)
+	: Phase2TrackerDigi::pixelToChannel(XtalkPrev.first, XtalkPrev.second);
+      signalNew.insert(std::pair<int,DigitizerUtility::Amplitude>(chanXtalkPrev, DigitizerUtility::Amplitude(signalInElectrons_Xtalk, nullptr, -1.0)));
```
Comment (kpedro88) [inline SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizerAlgorithm.cc]: `signalNew.emplace(chanXtalkPrev, DigitizerUtility::Amplitude(signalInElectrons_Xtalk, nullptr, -1.0));`

File: SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizerAlgorithm.cc [hunk 5]
```diff
@@ -592,62 +592,75 @@ void Phase2TrackerDigitizerAlgorithm::induce_signal(const PSimHit& hit,
...
+    }
+    if (hitChan.first < (numRows-1)) {
+      std::pair<int,int> XtalkNext = std::pair<int,int>(hitChan.first+1, hitChan.second);
+      int chanXtalkNext = (pixelFlag) ? PixelDigi::pixelToChannel(XtalkNext.first, XtalkNext.second)
+	: Phase2TrackerDigi::pixelToChannel(XtalkNext.first, XtalkNext.second);
+      signalNew.insert(std::pair<int,DigitizerUtility::Amplitude>(chanXtalkNext, DigitizerUtility::Amplitude(signalInElectrons_Xtalk, nullptr, -1.0)));
```
Comment (kpedro88) [inline SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizerAlgorithm.cc]: `signalNew.emplace(chanXtalkNext, DigitizerUtility::Amplitude(signalInElectrons_Xtalk, nullptr, -1.0));`

File: SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizerAlgorithm.cc [hunk 6]
```diff
@@ -592,62 +592,75 @@ void Phase2TrackerDigitizerAlgorithm::induce_signal(const PSimHit& hit,
...
+    }
+  }
+  for (auto const & l : signalNew) {
+    int chan = l.first;
+    auto iter = theSignal.find(chan);
+    if (iter != theSignal.end()) {
+      theSignal[chan] += l.second.ampl();
+    }  else {
+      theSignal.insert(std::pair<int,DigitizerUtility::Amplitude>(chan, DigitizerUtility::Amplitude(l.second.ampl(), nullptr, -1.0)));
```
Comment (kpedro88) [inline SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizerAlgorithm.cc]: `theSignal.emplace(chan, DigitizerUtility::Amplitude(l.second.ampl(), nullptr, -1.0));`

File: SimTracker/SiPhase2Digitizer/plugins/PixelDigitizerAlgorithm.cc [hunk 7]
```diff
@@ -117,3 +121,77 @@ void PixelDigitizerAlgorithm::accumulateSimHits(std::vector<PSimHit>::const_iter
...
+//  Add  Cross-talk contribution
+//
+// ======================================================================
+void PixelDigitizerAlgorithm::add_cross_talk(const Phase2TrackerGeomDetUnit* pixdet) {
+  if (!pixelFlag) return;
+
+  const Phase2TrackerTopology* topol = &pixdet->specificTopology();
+
+  // cross-talk calculation valid for the case of 25x100 pixels
+  if ( topol->pitch().first != 0.0025 || !topol->pitch().second != 0.0100 ) return;  // please check that units are really cm!
```
Comment (kpedro88) [inline SimTracker/SiPhase2Digitizer/plugins/PixelDigitizerAlgorithm.cc]: use named constants for magic numbers

File: SimTracker/SiPhase2Digitizer/plugins/PixelDigitizerAlgorithm.cc [hunk 8]
```diff
@@ -117,3 +121,77 @@ void PixelDigitizerAlgorithm::accumulateSimHits(std::vector<PSimHit>::const_iter
...
+  uint32_t detID = pixdet->geographicalId().rawId();
+  signal_map_type& theSignal = _signal[detID];
+  signal_map_type signalNew;
+  int numRows = topol->nrows();
+  int numColumns = topol->ncolumns();
+
+  for (auto & s : theSignal) {
+    float signalInElectrons = s.second.ampl();   // signal in electrons
+    std::pair<int,int> hitChan;
+    hitChan = PixelDigi::channelToPixel(s.first);
```
Comment (kpedro88) [inline SimTracker/SiPhase2Digitizer/plugins/PixelDigitizerAlgorithm.cc]: do this on previous line

File: SimTracker/SiPhase2Digitizer/plugins/PixelDigitizerAlgorithm.cc [hunk 9]
```diff
@@ -117,3 +121,77 @@ void PixelDigitizerAlgorithm::accumulateSimHits(std::vector<PSimHit>::const_iter
...
+    float signalInElectrons_odd_row_Xtalk_next_row = signalInElectrons * odd_row_interchannelCoupling_next_row; 
+    float signalInElectrons_even_row_Xtalk_next_row = signalInElectrons * even_row_interchannelCoupling_next_row;    
+    float signalInElectrons_odd_column_Xtalk_next_column = signalInElectrons * odd_column_interchannelCoupling_next_column;   
+    float signalInElectrons_even_column_Xtalk_next_column = signalInElectrons * even_column_interchannelCoupling_next_column;
+    
+    //subtract the charge which will be shared 
+    s.second.set(signalInElectrons-signalInElectrons_odd_row_Xtalk_next_row-signalInElectrons_even_row_Xtalk_next_row-signalInElectrons_odd_column_Xtalk_next_column-signalInElectrons_even_column_Xtalk_next_column);
+         
+    if (hitChan.first != 0) {	
+      std::pair<int,int> XtalkPrev = std::pair<int,int>(hitChan.first-1, hitChan.second);
```
Comment (kpedro88) [inline SimTracker/SiPhase2Digitizer/plugins/PixelDigitizerAlgorithm.cc]: same comments regarding use of `make_pair` and `emplace` from previous file

**Comment by kpedro88** [comment]
please test

**Comment by boudoul** [comment]
Hello @kpedro88 , do you know which AddOn test is failing ? strange that the bot is not telling ..

**Comment by kpedro88** [comment]
@boudoul I can't tell either. Let's relaunch the tests in case it was a transient xrootd issue or something

**Comment by kpedro88** [comment]
please test

**Comment by boudoul** [comment]
great this time the test passes  ! 
@civanch, @kpedro88, @mdhildreth  : As I mentioned in the ORP , It would be great if this PR could be considered for the next 10_4 it order to get as soon as possible relvals - thanks

**Comment by kpedro88** [comment]
+upgrade
comparison changes are isolated to 2023 workflows, and reflect 1) changes in pixel digitization, 2) changes in random number sequences throughout digitization

**Comment by fabiocos** [comment]
@civanch could you please comment or sign the PR in case it is ok for you?

**Comment by civanch** [comment]
+1

**Comment by fabiocos** [comment]
+1

**Pull request merged**